src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/TargetFrameworkProjectConfigurationDimensionProvider.cs

```diff
@@ -0,0 +1,84 @@
+        [ImportingConstructor]
+        public TargetFrameworkProjectConfigurationDimensionProvider(IProjectLockService projectLockService)
+        {
+            Requires.NotNull(projectLockService, nameof(projectLockService));
+
+            _projectLockService = projectLockService;
+        }
+
+        public async Task<IEnumerable<KeyValuePair<string, string>>> GetDefaultValuesForDimensionsAsync(UnconfiguredProject project)
+        {
+            // First target framework is the default one.
+            var targetFrameworks = await GetOrderedTargetFrameworksAsync(project).ConfigureAwait(false);
+            return ImmutableArray.Create(new KeyValuePair<string, string>(TargetFrameworkPropertyName, targetFrameworks.FirstOrDefault()));
+        }
+
+        public async Task<IEnumerable<KeyValuePair<string, IEnumerable<string>>>> GetProjectConfigurationDimensionsAsync(UnconfiguredProject project)
+        {
+            var targetFrameworks = await GetOrderedTargetFrameworksAsync(project).ConfigureAwait(false);
+            if (targetFrameworks.IsEmpty)
+            {
+                return ImmutableArray<KeyValuePair<string, IEnumerable<string>>>.Empty;
+            }
+
+            return ImmutableArray.Create(new KeyValuePair<string, IEnumerable<string>>(TargetFrameworkPropertyName, targetFrameworks));
+        }
+
+        private async Task<ImmutableArray<string>> GetOrderedTargetFrameworksAsync(UnconfiguredProject project)
+        {
+            Requires.NotNull(project, nameof(project));
+
+            using (var access = await _projectLockService.ReadLockAsync())
+            {
+                // Read the "TargetFrameworks" properties from msbuild project evaluation.
+                var configuredProject = await project.GetSuggestedConfiguredProjectAsync().ConfigureAwait(false);
+                var msbuildProject = await access.GetProjectAsync(configuredProject).ConfigureAwait(false);
+                var targetFrameworksStrings = msbuildProject.Properties.Where(p => p.Name.Equals(ConfigurationGeneral.TargetFrameworksProperty, StringComparison.OrdinalIgnoreCase));
+
+                if (!targetFrameworksStrings.Any())
+                {
+                    return ImmutableArray<string>.Empty;
+                }
+
+                // We need to ensure that we return the target frameworks in specified order.
+                var targetFrameworks = ImmutableArray.CreateBuilder<string>();
+                foreach (var frameworksString in targetFrameworksStrings)
+                {
+                    // TargetFrameworks contains semicolon delimited list of frameworks, for example "net45;netcoreapp1.0;netstandard1.4"
+                    foreach (var framework in frameworksString.EvaluatedValue.Split(';').Select(f => f.Trim()))
+                    {
+                        if (!string.IsNullOrEmpty(framework))
+                        {
+                            // CONSIDER: Do we need to do additional TFM validation here?
```

Comment:
No we don't need any validation - the idea is that the set of entries in the TargetFrameworks are opaque strings that we pass to the inner build.